pkg/container/versioned/value.go

```diff
@@ -0,0 +1,510 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package versioned
+
+import (
+	"errors"
+	"fmt"
+	"math"
+	"runtime"
+	"slices"
+	"strconv"
+	"sync/atomic"
+
+	"github.com/hashicorp/go-hclog"
+	"github.com/sirupsen/logrus"
+
+	"github.com/cilium/cilium/pkg/lock"
+	"github.com/cilium/cilium/pkg/logging/logfields"
+)
+
+type version uint64
+
+const (
+	// invalidVersion is never found from the value range.
+	// Also used as the upper bound for new values, so that the value is
+	// found when looking with 'maxVersion'
+	invalidVersion = version(math.MaxUint64)
+
+	// maxVersion in a VersionHandle finds the latest version of all non-removed values
+	maxVersion = version(math.MaxUint64 - 1)
+)
+
+// KeepVersion is an exported version type used when releasing memory held for old versions.
+type KeepVersion version
+
+var (
+	ErrInvalidVersion   = errors.New("invalid version")
+	ErrStaleTransaction = errors.New("stale transaction")
+	ErrStaleVersion     = errors.New("stale version")
+	ErrVersionNotFound  = errors.New("version not found")
+)
+
+// VersionHandle is used to keep values valid for a specific version from being released, so that
+// they are available for use for as long as the VersionHandle is not closed.
+//
+// A special form with a nil coordinator, which is returned by Latest(), always finds the latest
+// versions and does not keep any versions from being released.
+type VersionHandle struct {
+	// 'version' is the version this versionHandle keeps from being released
+	version version
+	// Coordinator of this versionHandle, if any. Used for closing if non-nil.
+	// Atomic due to nilling and for copy prevention.
+	coordinator atomic.Pointer[Coordinator]
+}
+
+func (h *VersionHandle) IsValid() bool {
+	return h != nil && (h.version == maxVersion || h.coordinator.Load() != nil)
+}
+
+func (h *VersionHandle) Version() KeepVersion {
+	return KeepVersion(h.version)
+}
+
+func (h *VersionHandle) String() string {
+	if h == nil {
+		return "version: <nil>"
+	}
+	validity := "invalid"
+	if h.IsValid() {
+		validity = "valid"
+	}
+	return fmt.Sprintf("%d (%s)", h.version, validity)
+}
+
+// Close releases the held version for removal once no handle for it are no longer held.
+// This may not be called while holding any locks that the 'closer' function passed to the
+// coordinator may take!
+func (h *VersionHandle) Close() (err error) {
+	if h == nil || h.version == invalidVersion {
+		return ErrInvalidVersion
+	}
+	// handle with 'maxVersion' is never closed
+	if h.version != maxVersion {
+		// Using CompareAndSwap makes sure each handle is closed at most once
+		coordinator := h.coordinator.Load()
+		if coordinator != nil && h.coordinator.CompareAndSwap(coordinator, nil) {
+			runtime.SetFinalizer(h, nil)
+			return coordinator.releaseVersion(h.version)
+		}
+		return ErrStaleVersion
+	}
+	return err
+}
```

Comment:
nit: let's try to avoid named returns if the type already sufficiently describes the return value, see https://go.dev/wiki/CodeReviewComments#named-result-parameters. `err` is never set anyway:

```suggestion
func (h *VersionHandle) Close() error {
	if h == nil || h.version == invalidVersion {
		return ErrInvalidVersion
	}
	// handle with 'maxVersion' is never closed
	if h.version != maxVersion {
		// Using CompareAndSwap makes sure each handle is closed at most once
		coordinator := h.coordinator.Load()
		if coordinator != nil && h.coordinator.CompareAndSwap(coordinator, nil) {
			runtime.SetFinalizer(h, nil)
			return coordinator.releaseVersion(h.version)
		}
		return ErrStaleVersion
	}
	return nil
}
```

Reply:
Done!